### PR TITLE
add precise roi pooling op test=develop

### PR DESCRIPTION
**Pull request by wopeizl**
The precise roi pooling implementation for paddle
please refer to https://arxiv.org/pdf/1807.11590.pdf

### CODE REVIEW

File: paddle/fluid/operators/prroi_pool_op.h [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+template <typename T>
+HOSTDEVICE T PrRoIPoolingMatCalculation(const T* this_data, const int s_h,
+                                        const int s_w, const int e_h,
+                                        const int e_w, const T y0, const T x0,
+                                        const T y1, const T x1, const int h0,
+                                        const int w0) {
+  T alpha, beta, lim_alpha, lim_beta, tmp;
+  T sum_out = 0;
+
+  alpha = x0 - T(s_w);
```
Comment (qingqing01) [inline paddle/fluid/operators/prroi_pool_op.h]: T() -> static_cast<T>(), same as below

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: Copyright (c) 2018  ->  Copyright (c) 2019

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+import numpy as np
+import unittest
+from op_test import OpTest
+
+
+class TestPRROIPoolOp(OpTest):
+    def set_data(self):
+        self.init_test_case()
+        self.make_rois()
+        self.calc_prroi_pool()
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: Maybe better to put core kenel `calc_prroi_pool` outside of `TestPRROIPoolOp`, make `TestPRROIPoolOp` simply and easy call `calc_prroi_pool` in other files by importint.

File: paddle/fluid/operators/prroi_pool_op.cu [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    GPUPRROIPoolForward<
+        T><<<blocks, threads, 0, ctx.cuda_device_context().stream()>>>(
+        output_size, in->data<T>(), rois->data<T>(), spatial_scale,
+        input_channels, height, width, output_channels, pooled_height,
+        pooled_width, rois_batch_id_list_gpu.data<int>(),
+        out->mutable_data<T>(ctx.GetPlace()));
+  }
+};
+
+template <typename Place, typename T>
```
Comment (heavengate) [inline paddle/fluid/operators/prroi_pool_op.cu]: Place -> DeviceContext， 或者这里应该可以删除这个typename

File: paddle/fluid/operators/prroi_pool_op.h [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+
+              output_data[output_index] = sum_out / win_size;
+            } else {
+              output_data[output_index] = 0.;
+            }
+          }
+        }
+      }
+    }
+    return;
```
Comment (heavengate) [inline paddle/fluid/operators/prroi_pool_op.h]: delete ？

File: paddle/fluid/operators/prroi_pool_op.h [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+        T sum_out = win_size == static_cast<T>(0.)
+                        ? static_cast<T>(0.)
+                        : *offset_output_grad_data / win_size;
+
+        int s_w = std::floor(win_start_w);
+        int e_w = std::ceil(win_end_w);
+        int s_h = std::floor(win_start_h);
+        int e_h = std::ceil(win_end_h);
+
+        // Accumubin_arealate diff_val into input data
```
Comment (heavengate) [inline paddle/fluid/operators/prroi_pool_op.h]: Accumubin_arealate ？

File: paddle/fluid/operators/prroi_pool_op.h (hunk 1 above)
Comment (wopeizl) [reply paddle/fluid/operators/prroi_pool_op.h]: ok

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py (hunk 2 above)
Comment (wopeizl) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: ok

File: python/paddle/fluid/tests/unittests/test_prroi_pool_op.py (hunk 3 above)
Comment (wopeizl) [reply python/paddle/fluid/tests/unittests/test_prroi_pool_op.py]: ok

File: paddle/fluid/operators/prroi_pool_op.cu (hunk 4 above)
Comment (wopeizl) [reply paddle/fluid/operators/prroi_pool_op.cu]: ok

File: paddle/fluid/operators/prroi_pool_op.h (hunk 5 above)
Comment (wopeizl) [reply paddle/fluid/operators/prroi_pool_op.h]: ok

File: paddle/fluid/operators/prroi_pool_op.h (hunk 6 above)
Comment (wopeizl) [reply paddle/fluid/operators/prroi_pool_op.h]: ok

**Review by qingqing01 — DISMISSED** [review]
LGTM. Need to ask @xsrobin to review doc.

**Review by heavengate — DISMISSED** [review]

**Review by wanghaoshuang — DISMISSED** [review]

**Review by wanghaoshuang — DISMISSED** [review]

**Review by qingqing01 — DISMISSED** [review]

**Review by heavengate — DISMISSED** [review]

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -12032,6 +12033,67 @@ def psroi_pool(input,
...
+def prroi_pool(input,
+               rois,
+               output_channels,
+               spatial_scale,
+               pooled_height,
+               pooled_width,
+               name=None):
+    """
+    ${comment}
+
```
Comment (raindrops2sea) [inline python/paddle/fluid/layers/nn.py:12650]: 补充一下这个OP API的整体功能说明，全称，以及参考论文来源等。

Comment (wopeizl) [reply python/paddle/fluid/layers/nn.py:12650]: ok

**Review by qingqing01 — DISMISSED** [review]

**Review by kuke — DISMISSED** [review]
LGTM

**Review by heavengate — DISMISSED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Review by wanghaoshuang — APPROVED** [review]

**Review by heavengate — APPROVED** [review]

**Pull request merged**